src/common/ui/code_extras/mode/javascript-mixed/javascript-mixed.js

```diff
@@ -0,0 +1,304 @@
+        jsState.tokenize = getTokenQuasi(stream);
+      }
+
+      return _forceJsModeToQuasi;
+    }());
+
+
+    function prepareReparseStringTemplateInLocalMode(stream, state) {
+      dbg('Entering local hml/css mode...');
+      // spit out beginning backtick as a token, and leave the rest of the text for local mode parsing
+      stream.backUp(stream.current().length - 1);
+
+      // workaround needed for 1-line string template,
+      // to ensure the ending backtick is parsed correctly.
+      forceJsModeToQuasi(stream, state.jsState);
+    }
+
+    function exitLocalModeAndTokenEndingBacktick(stream, state) {
+      dbg('Exiting local html/css mode...');
+      // parse the ending JS string template backtick in js mode
+      return jsMode.token(stream, state.jsState);
+    }
+
+    function tokenInLocalMode(modeToUse, stream, state) {
+      // parse the text in local mode
+      state.localMode = state.localMode || modeToUse;
+      state.localState = state.localState || CodeMirror.startState(state.localMode);
+      const style = state.localMode.token(stream, state.localState);
+      dbg(`  local ${modeToUse.name} mode token - `, stream.current());
+      return style;
+    }
+
+    function matchRule(rules, stream, state) {
+      for (const r of rules) {
+        if (r.curContext === (state.maybeLocalContext || '<start>')) {
+          // dbg('  rule:', r.curContext, r.matchFn.toString());
+          const matched = r.run(stream, state);
+          if (matched) { break; }
+        }
+      }
+    }
+
+    class Rule {
+      constructor(curContext, matchFn, nextContext, caseMatchedFn, caseNotMatchedFn) {
+        this.curContext = curContext;
+        this.matchFn = matchFn;
+        this.nextContext = nextContext;
+        this.caseMatchedFn = caseMatchedFn; // optional
+        this.caseNotMatchedFn = caseNotMatchedFn; // optional
+      }
+
+      run(stream, state) {
+        if (this.matchFn()) {
+          state.maybeLocalContext = this.nextContext;
+          if (state.maybeLocalContext == null) {
+            // local mode done, reset
+            state.localMode = null;
+            state.localState = null;
+          }
+          if (this.caseMatchedFn) { this.caseMatchedFn(); }
+          return true;
+        } // case rule transition criteria not matched
+        if (this.caseNotMatchedFn) {
+          this.caseNotMatchedFn();
+        } else { // default not matched logic: reset local mode matching
+          state.maybeLocalContext = null;
+        }
+        return false;
+      }
+    }
+
+    const cssMatcher = (function () {
+      const cssMode = CodeMirror.getMode(config, { name: 'css' });
+
+      function maybeToken(stream, state, jsTokStyle) {
+        let tokStyle = STYLE_PASS;
+        const tokTyp = state.jsState.lastType;
+        const tokStr = stream.current();
+
+        // define the transition rules to enter local CSS mode;
+        const rules = [
```

Comment:
This code will run lots of times, needlessly recreating the same object. You can create it one time at defineMode beginning.

Same for the other cases below.

Reply:
One difficulty in moving the rules outside the function is that the logic defined in the rules requires access to various parameters/variables passed. With the rules inside the function, the lambdas can simply access them locally. 

If the rules are moved outside the function, the lambdas need other ways to bind/access. 

One way is to pass them via some an object as a context to run the lambdas, but it'd make the lambdas more verbose. E.g.,

```javascript
          new Rule('css-2', (ctx) => ctx.tokTyp === 'quasi', // if it's a string template
            'css-in',
            (ctx) => prepareReparseStringTemplateInLocalMode(cssMode, ctx.stream, ctx.state)),

          new Rule('css-in', () => ctx.stream.peek() === '`', // if it hits ending backtick for string template
            null, // then exiting local css mode
            (ctx) => { ctx.tokStyle = exitLocalModeAndTokenEndingBacktick(ctx.stream, state); },
            (ctx) => { ctx.tokStyle = tokenInLocalMode(ctx.stream, ctx.state); }), // else stay in local mode
```

Existing rules for reference:
```javascript
          new Rule('css-2', () => tokTyp === 'quasi', // if it's a string template
            'css-in',
            () => prepareReparseStringTemplateInLocalMode(cssMode, stream, state)),

          new Rule('css-in', () => stream.peek() === '`', // if it hits ending backtick for string template
            null, // then exiting local css mode
            () => { tokStyle = exitLocalModeAndTokenEndingBacktick(stream, state); },
            () => { tokStyle = tokenInLocalMode(stream, state); }), // else stay in local mode
```

Thoughts? Any other ideas?

Reply:
I'd say pass this stuff in `state`.

Reply:
Er, I mean your `ctx` variant is fine.

Reply:
Hmm, I'd probably even remove `tokStr` and just use `ctx.stream.current()` because not all rules need it...

Reply:
done

Reply:
done